flow/record/base.py

```diff
@@ -723,66 +789,73 @@ def open_path(path, mode, clobber=True):
             fp = io.open(path, mode)
         # check if we are reading a compressed stream
         if not out and binary:
-            if not hasattr(fp, "peek"):
-                fp = Peekable(fp)
-            peek_data = fp.peek(4)
-            if peek_data[:2] == GZIP_MAGIC:
-                fp = gzip.GzipFile(fileobj=fp, mode=mode)
-            elif HAS_BZ2 and peek_data[:3] == BZ2_MAGIC:
-                fp = bz2.BZ2File(fp, mode=mode)
-            elif HAS_LZ4 and peek_data[:4] == LZ4_MAGIC:
-                fp = lz4.open(fp, mode=mode)
-            elif HAS_ZSTD and peek_data[:4] == ZSTD_MAGIC:
-                dctx = zstd.ZstdDecompressor()
-                fp = dctx.stream_reader(fp)
+            fp = open_stream(fp, mode)
     return fp
 
 
-def RecordAdapter(url, out, selector=None, clobber=True, **kwargs):
-    url = str(url or "")
-
+def RecordAdapter(url, out: bool, selector=None, clobber=True, **kwargs) -> AbstractWriter | AbstractReader:
     # Guess adapter based on extension
     ext_to_adapter = {
         ".avro": "avro",
         ".json": "jsonfile",
         ".jsonl": "jsonfile",
         ".csv": "csvfile",
     }
-    _, ext = os.path.splitext(url)
+    cls_stream = None
+    if not out and hasattr(url, "read") and hasattr(url, "seek"):
+        # This record adapter has received a file-like object.
+        # We just need to find the right adapter
+        # by peeking into the first few bytes.
 
-    adapter_scheme = ext_to_adapter.get(ext, "stream")
-    if "://" not in url:
-        url = f"{adapter_scheme}://{url}"
+        # First, we open the stream. If the stream is compressed, open_stream will wrap it for us into a decompressor.
+        cls_stream = Peekable(open_stream(url, "rb"))
 
-    p = urlparse(url, scheme=adapter_scheme)
-    adapter, _, sub_adapter = p.scheme.partition("+")
+        # We will again peek into the bytes to find an adapter, so we wrap the cls_stream into a Peekable object.
+        adapter = find_adapter_for_stream(cls_stream)
```

Comment:
Would it be possible to move the `Peekable()` wrapper into `find_adapter_for_stream()`, like what you do in `open_stream()`.